PS3Eye Native CameraType v0.1

ifly53e opened the pull request:
This PR takes into account commit #576

Changes to main is to call the PS3EyeCamera init() and changes to CameraManager are to call the PS3EyeCamera config window.

Follow the instructions to install the proper PS3Eye usb driver from this link:
https://github.com/cboulay/psmove-ue4/wiki/Windows-PSEye-Setup

You will need a program called Zadig to help install the usb driver:
http://zadig.akeo.ie/downloads/zadig_2.2.exe

Missing from the Zadig instructions is to click on the Options menu and click "List All Devices" so that you can see the PS3Eye camera in the first place.

Test your setup with ps3eye_sdl.exe found at:
https://github.com/cboulay/psmove-ue4/tree/master/Binaries/Win64

Start ShootOFF and it should see the PS3Eye.

Auto exposure adjustment works during autocalibration.  

You can also right click on the feed in debug mode to bring up the configure camera menu and adjust the gain and exposure.  You can also toggle "auto gain" on and off and see the current FPS there.

ifly53e committed:
changes to ps3eye with new branch

- removed dead code
- added getCameraState
- changed bytebuffer to byte[] so no conversion is required
- shortened FPS in config to two decimals

ifly53e committed:
Revert "changes to ps3eye with new branch"

This reverts commit 5f1902c9fc5d79be4c123ad49176d54a9c3ed929.

ifly53e committed:
Proper commit for pull request

-removed dead code
-added getState()
-changed bytebuffer to byte[] (what native coded wanted anyway)
-fixed fps output in config window to two decimal places
-does not include shootoff.properties in the commit

setuidloot commented on src/main/java/com/shootoff/Main.java:
```diff
@@ -680,6 +681,7 @@ public static void closeNoV4lCompat(File v4lCompat) {
 	@Override
 	public void start(Stage primaryStage) {
 		OptiTrackCamera.init();
+		PS3EyeCamera.init();
```
This change is fine.

ifly53e replied:
woohoo!

setuidloot commented on src/main/java/com/shootoff/camera/CameraManager.java:
```diff
@@ -766,6 +767,9 @@ public void launchCameraSettings() {
 		if (camera instanceof SarxosCaptureCamera) {
 			((SarxosCaptureCamera) camera).launchCameraSettings();
 		}
+		if (camera instanceof PS3EyeCamera) {
```
For clarity for additional future cameras, please make this an "else if"

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+	protected Optional<CameraEventListener> cameraEventListener = Optional.empty();
+	protected long currentFrameTimestamp = -1;
+	private Dimension dimension = null;
+	private int viewWidth = 0;
```
Since "size_x" and "size_y" are used to open the camera anyway, please set viewWidth and viewHeight to 640x480 and then use them throughout the file instead of "size_x" and size_y".  Make them static and final and make getViewWidth/Height use them as well.

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+	private long lastCameraTimestamp = -1;
+	private long lastFrameCount = 0;
+
+	static eyecam eyecamLib = (eyecam) Native.loadLibrary("eyecam", eyecam.class);
```
My understanding is that this will throw an exception if the dll is not present, which will not work on Linux.  This should be a dynamic call that exits if the module can't be loaded.  See OptiTrackCamera for an example of failing safe on "UnsatisfiedLinkError".

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+		if (initialized)
+			return;
+
+		eyecamLib.ps3eye_init();
```
This is where the code needs to be safe for if the DLL is not present.

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+		eyecamLib.ps3eye_init();
+
+		if (eyecamLib.ps3eye_count_connected() == 1) {
+			logger.debug("Opening ps3eye");
```
Please change the logger levels to Trace except for cases where a layman might find use in the message.  For example, "ps3eye dll successfully loaded"  "ps3 camera found and registered" would be acceptable debug level messages.

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+		return ps3FPS;
+	}
+
+	protected void setFPS(double newFPS) {
```
The FPS and timestamp code looks the same as CalculatedFPSCamera, is there a reason it wasn't extended?

ifly53e replied:
I extended calculatedFPSCamera and took out a lot of the functions that were duplicated.  I started from OptiTrackCamera so that is why it was not extended.

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+	@Override
+	public ShotDetector getPreferredShotDetector(final CameraManager cameraManager, final Configuration config,
+			final CameraView cameraView) {
+		if (OptiTrackShotDetector.isSystemSupported())
```
This has to be removed, it will not work with this camera.

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+		final int newExp = (int) (curExp - (.1 * (double) curExp));
+		logger.debug("curExp[ {} newExp {}", curExp, newExp);
+
+		if (newExp < 20)
```
Have you checked to see what the minimal reasonable value for this is for the ps3eye? Unlikely to be the same as the optitrack.  This is the value at which the autocalibration should stop trying to reduce it.  Ultimately it should be in a static variable at the top of the class as well (I will be changing this in optitrack at some point).

ifly53e replied:
changed this to 17 after playing with the adjustment a bit

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+			return false;
+
+		setExposure(newExp);
+		logger.debug("curExp[ {} newExp {} res {}", curExp, newExp, getExposure());
```
Also should be at trace.  The reason I have more scrutiny for your code than mine (Optitrack) is that I'm the only person with an optitrack, so I am not concerned with causing any users problems.  The ps3eye is pretty common, though.

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+			setExposure(origExposure.get());
+	}
+
+	public boolean limitsFrames() {
```
This should be false, the ps3eye yields every frame.  Sorry this isn't documented yet.

ifly53e replied:
done

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -0,0 +1,621 @@
+//written by ifly53e
```
The file has to have the ShootOFF license.  You can include your name before or after the license stating that you wrote the file originally.

ifly53e replied:
done.  I also added the install instructions as comments

cbdmaul reviewed (changes requested):
Needs some refinement but it is not far from being accepted.  Please point us at the source repo for the eyeCam dll.  We need to review the license and possibly compile it ourselves to maintain our licensing correctly.

ifly53e committed:
Merge remote-tracking branch 'refs/remotes/phrack/master' into PS3Eye

ifly53e committed:
Merge remote-tracking branch 'refs/remotes/phrack/master' into PS3Eye

ifly53e committed:
Requested changes made

Made the requested changes by cdmaul.
PS3EyeCamera extends CalculatedFPSCamera

ifly53e commented:
The source for the driver is at https://github.com/inspirit/PS3EYEDriver.  I added a small test function and I compiled eyecam.dll.  They don't release any dll so I had to make my own.

I believe I made all of your requested changes so if I missed something, let me know and I will make the edits.

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -77,18 +111,28 @@ public static void init() {
+			initialized = false;
+		}
+
 		eyecamLib.ps3eye_init();
```
eyecamLib will be null if there is an UnsatisfiedLinkError.  The function must end before this point if so.

ifly53e replied:
Two choices here.  I either put a return at the end of the catch block, or I put all of the init code in the try block.  Which is better?

setuidloot replied:
I think the way you had it before was better, with the following changes:
1. Return after setting initalized false in the catch block
2. Before calling eyecamLib for the first time, make sure it is not null.  If it is null, set initialized false and return.  This is the safest way to handle the possibility that eyecamLib is not what it is supposed to be.  Your current code would not catch a null eyecamLib not resulting from a UnsatisfiedLinkError

setuidloot commented on src/main/java/com/shootoff/camera/cameratypes/PS3EyeCamera.java:
```diff
@@ -613,7 +577,7 @@ protected ps3eye_t(Pointer ps3eye_t) {
 		int ps3eye_get_parameter(ps3eye_t eye, int ps3eyeGain);
 
-		// a test returns 85
+		// a test inserted by ifly53e returns 85
```
Since this isn't used, and presumably you no longer need it for testing, please remove this function definition.

ifly53e replied:
done

cbdmaul commented:
Phrack will have to review the licensing, but that doesn't look like a problem.

ifly53e committed:
PR changes

removed function definition not used anymore
resolved try catch block logic error

ifly53e committed:
fixed eyecamLib possible null issue

Took code out of try and added return to catch

Added check for eyecamLib being null not resulting fron an
UnsatisifiedLinkError

phrack commented:
Nice work guys!

ifly53e commented:
Thanks.  Was the dll supposed to be included in the nightly build?  I
couldn't find it in the latest download just now.  The dll has to be in the
root directory for it to work.

On Sep 21, 2016 3:09 PM, "phrack" notifications@github.com wrote:

Nice work guys!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
https://github.com/phrack/ShootOFF/pull/578#issuecomment-248757643, or mute
the thread
https://github.com/notifications/unsubscribe-auth/ACr7fkEOYTS67ce5GxXeYlGSvToSsWHkks5qsasFgaJpZM4KCOXe
.

phrack commented:
We need a copy task for it in build.gradle and the fxJar task should depend on it. There are enough examples of doing this it should be easy to submit a PR to do it. Search for `task fxJar(dependsOn: build) {` in build.gradle.

phrack commented:
Fix the fact that PS3EyeCamera.cameraClosed() is never used while you're at it please.
